jupyterhub/events.py

```diff
@@ -0,0 +1,131 @@
+"""
+Emit structured, discrete events when various actions happen.
+"""
+from traitlets.config import Configurable
+
+import logging
+from datetime import datetime
+import jsonschema
+from pythonjsonlogger import jsonlogger
+from traitlets import TraitType, List
+import json
+import six
+
+
+class Callable(TraitType):
+    """
+    A trait which is callable.
+
+    Classes are callable, as are instances
+    with a __call__() method.
+    """
+    info_text = 'a callable'
+    def validate(self, obj, value):
+        if six.callable(value):
+            return value
+        else:
+            self.error(obj, value)
+
+def _skip_message(record, **kwargs):
+    """
+    Remove 'message' from log record.
+
+    It is always emitted with 'null', and we do not want it,
+    since we are always emitting events only
+    """
+    del record['message']
+    return json.dumps(record, **kwargs)
+
+
+class EventLog(Configurable):
+    """
+    Send structured events to a logging sink
+    """
+    handlers_maker = Callable(
```

Comment:
See my comments [here](https://github.com/jupyterlab/jupyterlab-telemetry/pull/1/files#r294020974). 

Wondering if we can make change this trait to be a list of "event sinks". As it stands, I think this trait puts too much work on the shoulders of admins. Maybe we can create a standard "EventSink" interface. Admins can use this interface to hook their event sink service into Jupyterhub's event-logger by creating an EventSink for their service. Then, they just list it in the `event_sinks` trait.

Reply:
Yeah, I agree that a callable isn't a viable solution. It's here as a hack copied from binderhub due to a limitation of traitlest - you can only set traitlets to values that are pickleable! This means most objects containing a Logger instance can't be set to traitlet values in any form. We should find a way to work around that - requiring admins to pass in a callable isn't an interface we should ship with.

What would our own custom EventSink class provide that the python Logger class does not? If we have our own layer, someone needs to write a wrapper (and documentation) for every logging service out there. This is already written for the standard python logging module, so unless we have very strong reasons we should just stick to that IMO...